src/library_lz4.js

```diff
@@ -0,0 +1,183 @@
+// TODO: put behind a flag
+mergeInto(LibraryManager.library, {
+  $LZ4__deps: ['$FS'],
+  $LZ4: {
+    DIR_MODE: {{{ cDefine('S_IFDIR') }}} | 511 /* 0777 */,
+    FILE_MODE: {{{ cDefine('S_IFREG') }}} | 511 /* 0777 */,
+    CHUNK_SIZE: -1,
+    codec: null,
+    init: function() {
+      if (LZ4.codec) return;
+      LZ4.codec = (function() {
+        {{{ read('mini-lz4.js') }}};
+        return MiniLZ4;
+      })();
+      LZ4.CHUNK_SIZE = LZ4.codec.CHUNK_SIZE;
+    },
+    loadPackage: function (pack) {
+      LZ4.init();
+      var compressedData = pack['compressedData'];
+      if (!compressedData) compressedData = LZ4.codec.compressPackage(pack['data']);
+      assert(compressedData.cachedIndexes.length === compressedData.cachedChunks.length);
+      for (var i = 0; i < compressedData.cachedIndexes.length; i++) {
+        compressedData.cachedIndexes[i] = -1;
+        compressedData.cachedChunks[i] = compressedData.data.subarray(compressedData.cachedOffset + i*LZ4.CHUNK_SIZE,
+                                                                      compressedData.cachedOffset + (i+1)*LZ4.CHUNK_SIZE);
+        assert(compressedData.cachedChunks[i].length === LZ4.CHUNK_SIZE);
+      }
+      console.log('loading package');
+      pack['metadata'].files.forEach(function(file) {
+        var dir = PATH.dirname(file.filename);
+        var name = PATH.basename(file.filename);
+        FS.ensureFolder(dir, true, true);
+        var parent = FS.analyzePath(dir).object;
+        LZ4.createNode(parent, name, LZ4.FILE_MODE, 0, {
+          compressedData: compressedData,
+          start: file.start,
+          end: file.end,
+        });
+      });
+    },
+    createNode: function (parent, name, mode, dev, contents, mtime) {
+      var node = FS.createNode(parent, name, mode);
+      node.mode = mode;
+      node.node_ops = LZ4.node_ops;
+      node.stream_ops = LZ4.stream_ops;
+      node.timestamp = (mtime || new Date).getTime();
+      assert(LZ4.FILE_MODE !== LZ4.DIR_MODE);
+      if (mode === LZ4.FILE_MODE) {
+        node.size = contents.end - contents.start;
+        node.contents = contents;
+      } else {
+        node.size = 4096;
+        node.contents = {};
+      }
+      if (parent) {
+        parent.contents[name] = node;
+      }
+      return node;
+    },
+    node_ops: {
+      getattr: function(node) {
+        return {
+          dev: 1,
+          ino: undefined,
+          mode: node.mode,
+          nlink: 1,
+          uid: 0,
+          gid: 0,
+          rdev: undefined,
+          size: node.size,
+          atime: new Date(node.timestamp),
+          mtime: new Date(node.timestamp),
+          ctime: new Date(node.timestamp),
+          blksize: 4096,
```

Comment:
Does it make sense to configure block size? Or make it constant to avoid repetition? Though I see it's only referenced in these two lines I suppose..